aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java

```diff
@@ -88,6 +91,32 @@ private static AwsClientFactory loadClientFactory(String impl, Map<String, Strin
     return factory;
   }
 
+  /**
+   * Metrics are the only reliable way provided by the AWS SDK to determine if an API call was
+   * retried. Not for public use.
+   */
+  public static class InternalRetryDetector implements MetricPublisher {
+    private boolean retried = false;
```

Comment:
I remember that retries could happen in parallel. Should we make this an AtomicBoolean?

Reply:
I don't believe parallel retries are done by the non-async client, but on the off chance you are right, or become right in the future, it doesn't hurt to make this an `AtomicBoolean`, so will change.